Python Script for JSON Config Files

catherinemeadows opened the pull request:
Related to issue #184. Created a python script that parses a JSON config file to start up a chain of NFs. 

## Summary:
This feature makes it easier for the user to launch a chain of NFs by running a python script that parses a JSON Config file containing the desired NFs and parameters. To run the script (within the /examples folder): ```python config.py [JSON Config file name]```.  Normal shutdown of NFs occurs on user termination (ctrl+c). Shutdown of NFs also occurs if any NF terminates unexpectedly or fails to start up. 

Documentation for using this feature was added [here](https://github.com/catherinemeadows/openNetVM/blob/configScript/docs/Examples.md) and [here](https://github.com/catherinemeadows/openNetVM/blob/configScript/docs/NF_Dev.md). 

**Usage:**

<!-- Check list of the things this PR accomplishes -->
| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          | <!-- Provide a list of issues --> 
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   | 👍   
| Bug fixes                |
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 


<!-- If the pr has any dependencies or merge quirks note them here -->
## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [X] PR is ready for review


<!-- What you did to test the PR, what needs to be done -->
## Test Plan:
To test, I ran the python script to start up chains of NFs. Screenshots from running the script on the [example_chain.json](https://github.com/catherinemeadows/openNetVM/blob/configScript/examples/example_chain.json) file in /examples:

![test_1](https://user-images.githubusercontent.com/43352540/83726668-debffb80-a611-11ea-9090-ed81dcb584d2.png)

While running, processes are killed on user shutdown (Ctrl + C): 

![user_exit](https://user-images.githubusercontent.com/43352540/83726745-ff885100-a611-11ea-9de6-c63bee8c762b.png)

I also tested the script in the case that all NFs do not start (i.e. one or more NF has some error). The script shuts down all processes in this case and outputs the error from the failed NF(s). For this test case, the same [config file](https://github.com/catherinemeadows/openNetVM/blob/configScript/examples/example_chain.json) was used but with a '-d' flag purposefully omitted from the second set of parameters for the speed_tester. 
![error_exit](https://user-images.githubusercontent.com/43352540/83727002-5ee66100-a612-11ea-85d2-a1cdf641bdf9.png)

onvm commented:
> In response to PR creation

## CI Message

Your results will arrive shortly

bdevierno1 commented:
[Information](https://docs.pylint.org/en/1.6.0/tutorial.html) about docstrings. Just a short comment about the overall functionality of your program and each function.

catherinemeadows commented:
Not sure what to do about the two linter errors about unused arguments for the handler. That function just listens/catches SIGINT (CTRL -C) to shut down the NFs and the script.

kevindweb commented on examples/config.py:
```diff
@@ -0,0 +1,138 @@
+    for k, v in data.items():
+        for item in v:
+            nf_list.append(k)
+            PARAM = str(item).strip("'{[]}'")
+            PARAM = remove_prefix(PARAM, "u'parameters': u'")
```
(goes along with my strip pylint comment). What's the PARAM argument doing? Can we do something like 
```py
cmds_list.append("./go.sh " v.parameter)
``` 
instead? to leverage the JSON syntax?

catherinemeadows replied:
So, I worked with doing something like v.parameter or item.parameter, but was having some issues dealing with encoding/decoding and then it not letting me concat unless item was a command. I just did some more research though, and I found a better way to do this. I'll push my edits in a few minutes!

kevindweb replied:
Awesome, great work!

kevindweb commented:
For your `handler` function I would follow this [ignore pylint error](https://stackoverflow.com/questions/28829236/is-it-possible-to-ignore-one-single-specific-line-with-pylint). 

For 
```
examples/config.py:120:0: C0301: Line too long (108/100) (line-too-long)
```
you can always write a `\` to split lines (even strings) onto [multiple lines](https://stackoverflow.com/a/24331604/10158041). 

For your stripping issue, I don't know what the split is doing, but there might be a way to optimize it

catherinemeadows commented on examples/config.py:
```diff
@@ -0,0 +1,99 @@
+                out, err = p.communicate()
+                on_failure(out + err)
+                break
+            if ret_code is not None:
```
Good catch! line 97 used to just be an "else". I think something weird happened with a merge at some point. p.communicate() gives the output from the NF. I can change that to only print out the error.

catherinemeadows commented on examples/config.py:
```diff
@@ -0,0 +1,99 @@
+            pass
+    for n in nf_list:
+        try:
+            os.system("sudo pkill " + n)
```
Before, when I was using multiprocessing, I had to kill the NF as well as kill the python processes or else they would keep running in the background, but I just did some tests and it seems like I don't need that when using Popen, so I can take that out. Thank you for reviews!

onvm commented:
> Don't fail test

## CI Message

Your results will arrive shortly

catherinemeadows commented:
Now, a user can specify a directory name in command line (otherwise, default dir name is just the timestamp) so a directory is made, containing a .txt log file for each launched NF (name of log files follow format "log-NF name-instance ID", ex: log-speed_tester-1.txt). Output from NFs now continuously update their corresponding .txt files. User can track output of a particular NF with ```tail -f log file```. 

To run (within examples folder): ```python config.py <config file name> [desired dir name]```

Screenshot below shows the output of running ```python config.py example_chain.json``` and then in a separate pane, within the timestamp directory that was created, running ```tail -f log-simple_forward-2.txt``` 

![log_file](https://user-images.githubusercontent.com/43352540/84360949-ec88fa00-ab98-11ea-98ab-8874db02b6ff.png)

kevindweb commented:
Great addition, will be super helpful for debugging purposes!

EthanBaron14 commented on examples/config.py:
```diff
@@ -0,0 +1,148 @@
+            ret_code = pl.poll()
+            if ret_code is not None:
+                out, err = pl.communicate()
+                on_failure(out + err)
```
Getting a weird error on this line. To reproduce, run the config file and exit the manager with Control + C. 
```python
Traceback (most recent call last):
  File "config.py", line 148, in <module>
    running_services()
  File "config.py", line 91, in running_services
    on_failure(out + err)
TypeError: unsupported operand type(s) for +: 'NoneType' and 'NoneType'
```
I'm also seeing this error when I try to run the config file without the manager running. It creates the log folders and logs and populates each log with the information that the manager is not running, but still crashes with this error.

Also seeing this when I duplicate one of the json entries. An example of this would be copy/pasting the speed_tester NF chain:
```json
        "speed_tester": [
                {
                        "parameters": "1 -d 2 -c 16000"
                },
		{
			"parameters": "3 -d 3 -c 16000"
                }
        ],
        "speed_tester": [
                {
                        "parameters": "1 -d 2 -c 16000"
                },
		{
			"parameters": "3 -d 3 -c 16000"
                }
        ]
```
Also seeing this when I duplicate one of the chain calls:
```json
        "speed_tester": [
                {
                        "parameters": "1 -d 2 -c 16000"
                },
		{
			"parameters": "3 -d 3 -c 16000"
                },
                {
			"parameters": "3 -d 3 -c 16000"
                }
        ]
```

catherinemeadows replied:
I'm not getting the first error that you mentioned when I exit with Ctrl + C, but I'll look more into this. It seems for error handling, something broke when I added the logging so I'll fix that

EthanBaron14 commented on examples/config.py:
```diff
@@ -0,0 +1,148 @@
+    try:
+        os.mkdir(log_dir)
+    except OSError:
+        on_failure("Creation of directory %s failed" % (dir))
```
`dir` appears to be a built-in Python function. I think `(dir)` is supposed to be `(log_dir)` here, as you're just calling that method and it's returning standard output from that command:
```
Creation of directory <built-in function dir> failed
Error occurred. Exiting...
```
To verify this theory, open Python 3 and run the `dir` command:
```
$ python3
Python 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> dir
<built-in function dir>
```

Furthermore, it seems like it's requiring that we allow the script to create a new directory for the log files. While I understand the motivation for this because it'll alleviate the possibility of using a non-empty directory (which could be messy), the error message is a little vague. By saying that it failed to create the directory, it could mean several things, like no permission, a directory already exists, etc.. I tested this by changing the `examples` folder's permissions to prohibit the script from creating the folder and by also trying to run the script in a situation where the directory already exists. Alternatively, we could also allow the script to use a directory that was just created, like I did. Just something to consider, but the main issue here is calling `dir` instead of using `log_dir`. Additionally, if we just wanted to use one directory and run this multiple times, we would be forced to delete the one that already exists and start over.

catherinemeadows replied:
Yep, that was a typo. I meant log_dir, thanks!

EthanBaron14 commented on examples/config.py:
```diff
@@ -0,0 +1,148 @@
+    config_file = get_config()
+
+    with open(config_file) as f:
+        data = json.load(f)
```
If the json is improperly formatted, we get an error here:
```python
Traceback (most recent call last):
  File "config.py", line 119, in <module>
    data = json.load(f)
  File "/usr/lib/python3.6/json/__init__.py", line 299, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.6/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 15 column 1 (char 311)
```

Not sure if this is something you really want to or need to fix, since the user should be using correctly formatted json, but something to consider nonetheless. That error was caused by including an extra comma after the closing `]` for speed_tester

catherinemeadows replied:
We definitely want it to print an error statement if the JSON file has incorrect syntax. It used to do this, but something weird happened when I added the log files, so I'll look into this

catherinemeadows commented:
@EthanBaron14 when I switched the processes to log the output to files instead of PIPE, it created an error because the error messages couldn't be printed anymore. I fixed this, but do you mind testing again? I wasn't getting a few of the errors that you mentioned in the first place, so I just want to be sure. Specifically, I don't see special characters in the log files and it was exiting fine on Ctrl + C for me, but I can look more into that if you're still getting those errors. 

Those fixes have been pushed, but I'll also add functionality for specifying a directory that already exists and some more print statements for when the NFs are starting, etc. Thanks for your review!

kevindweb commented:
> I really like what you're doing here. Just a few things I've noticed...
> 
> The logging seems to be a little wonky. Not sure what might be causing it, but the log files are changing some characters around. This is an excerpt from the simple_forward log that got generated when running your default configuration:
> 
> ```
> No IP4 header found
> �[2J�[1;1HPACKETS
> -----
> Port : 0
> Size : 14
> N°   : 388000000
> ```
> 
> Both instances of the speed_tester NF have similar issues with logging.
> 
> I would personally maybe have one or two messages on the stdout end like "Creating directory", "Starting NFs", "NFs successfully started", or "Shutting down NFs" but that's just personal preference.
> 
> In a perfect situation, though, this script runs perfectly. Really, really nice work.

@EthanBaron14 `�[2J�[1;1HPACKETS` these are terminal formatters that you're seeing in the logfile. If you run `cat logfile` instead, it will print out normally.

EthanBaron14 commented:
> > I really like what you're doing here. Just a few things I've noticed...
> > The logging seems to be a little wonky. Not sure what might be causing it, but the log files are changing some characters around. This is an excerpt from the simple_forward log that got generated when running your default configuration:
> > ```
> > No IP4 header found
> > �[2J�[1;1HPACKETS
> > -----
> > Port : 0
> > Size : 14
> > N°   : 388000000
> > ```
> > 
> > 
> > Both instances of the speed_tester NF have similar issues with logging.
> > I would personally maybe have one or two messages on the stdout end like "Creating directory", "Starting NFs", "NFs successfully started", or "Shutting down NFs" but that's just personal preference.
> > In a perfect situation, though, this script runs perfectly. Really, really nice work.
> 
> @EthanBaron14 `�[2J�[1;1HPACKETS` these are terminal formatters that you're seeing in the logfile. If you run `cat logfile` instead, it will print out normally.

Looks like you're right. When I try to view it in any text editor, though, it still has those formatters. Makes it so the only way to view the log is with `cat`, right?

kevindweb commented:
Well you can just avoid the formatters when you read it, or we could write something to dynamically remove the formatters before the logfile?

catherinemeadows commented:
@EthanBaron14 added a try/except block for loading the JSON file for the case that the JSON config has invalid syntax

kevindweb commented on examples/config.py:
```diff
@@ -0,0 +1,159 @@
+from subprocess import Popen
+from datetime import datetime
+
+def handler(signal_received, frame):
+    """Handles cleanup on user shutdown"""
+    # pylint: disable=unused-argument
```
Sorry for the nit but I knew this was a thing in Python https://stackoverflow.com/a/54213922/10158041

```suggestion
def handler(signal_received, _):
    """Handles cleanup on user shutdown"""
```

catherinemeadows replied:
When I make this fix, I'm still getting a linter error for unused argument "signal_received". If I change both arguments to "_", I get linter errors for duplicate argument names. Not sure if there's a fix for this?

kevindweb commented:
<details>
<summary>How do I dropdown?</summary>
<br>
This is how you dropdown.
<br><br>
<pre>
<details>
<summary>Test?</summary>
<pre>
will this 
</details>
</pre>
</details>

kevindweb commented on examples/config.py:
```diff
@@ -0,0 +1,159 @@
+            ret_code = pl.poll()
+            if ret_code is not None:
+                on_failure()
+                break
+            time.sleep(.1)
+            continue
```
I think these lines should be updated because the break and continue are redundant. @catherinemeadows please correct me, but I think the sleep should be unindented (outside the for loop) so we can catch the errors between NFs faster?
```suggestion

        time.sleep(.1)
```

On a side note, do you think it's possible to run a timer in this function and allow the user to optionally tell config.py how many seconds to run and die automatically without `^C`? This would help with automating the script.

kevindweb commented:
@onvm quick testing

onvm commented:
> @onvm quick testing

## CI Message

Your results will arrive shortly

onvm commented:
> @onvm quick testing

## CI Message

Error: Failed to parse nimbnode23 stats

kevindweb commented:
@onvm again pls

onvm commented:
> @onvm again pls

## CI Message

Your results will arrive shortly

WilliamMaa commented:
@catherinemeadows on line 190, can you change the print to print("Starting %s %s" % (nf, cmd), flush=True). Sorry but I need this to make the python server be able to properly close the nf chain.

dennisafa commented:
This looks good to go from my end but I'll let @WilliamMaa make the call since you were assigned

twood02 commented:
@catherinemeadows and @WilliamMaa will discuss the NF shutdown issue today.  @catherinemeadows will post example JSON file

catherinemeadows commented:
example JSON config file (this is the one will be included in the /examples folder: 
```
{
        "globals": [
                {
                        "TTL": 10
                },
                {
                        "directory": "output_files"
                }
        ],
        "simple_forward": [
                {
                        "parameters": "2 -d 1"
                }
        ],
        "speed_tester": [
                {
                        "parameters": "1 -d 2 -c 16000"
                },
		{
			"parameters": "3 -d 3 -c 16000"
                }
        ]
}
```

twood02 commented:
Are there any other global parameters and if so are they defined in the Readme?

Could you easily add a `directory-prefix` option, which would let you give a dir name which would then be appended with the date/time? This would be helpful for organizing different experiment runs.

catherinemeadows commented:
Those are the only two globals I have now. I can add a directory-prefix option. I wrote documentation for the globals and how to run the script here: https://github.com/catherinemeadows/openNetVM/blob/configScript/docs/NF_Dev.md

catherinemeadows commented:
@WilliamMaa that was just a git mess up on my end. I'll get rid of those changes

twood02 commented on docs/NF_Dev.md:
````diff
@@ -99,17 +103,58 @@ values.
 ```
 
+Additionally, a developer can run the [config python script](../examples/config.py) to deploy multiple network functions, including linear or circular chains of multiple NFs, from a JSON config file. An example config file can be found [here](../examples/example_nf_deploy.json). In the config file, a user can specify a number of seconds for the NFs to run before shutdown with the global "TTL". If ...
+When the NFs are launched, if the user specifies a directory name with the global "directory", a directory will be created (if it does not already exist) with the user-specified name. If a global "directory-prefix" is specified, a directory will be created with the specified prefix + timestamp. If no directory name or prefix is specified, the default name of the created directory will the be the  ...
````
This paragraph is pretty dense. Can you split it up?  Or give a paragraph description and then a few bullet points explaining specific details?

One important point not mentioned is that the "NF Name" must match the name of a folder in the examples directory (right?)

twood02 commented on docs/NF_Dev.md:
````diff
@@ -99,17 +103,58 @@ values.
 }
 ```
 
+Additionally, a developer can run the [config python script](../examples/config.py) to deploy multiple network functions, including linear or circular chains of multiple NFs, from a JSON config file. An example config file can be found [here](../examples/example_nf_deploy.json). In the config file, a user can specify a number of seconds for the NFs to run before shutdown with the global "TTL". If ...
````
Put a heading above this like `## Running Groups of NFs` since this isn't really that tightly connected to the config file library (totally different other than using json, in fact)

twood02 commented on docs/NF_Dev.md:
````diff
@@ -99,17 +103,58 @@ values.
+
+```
+python3 config.py <config file name>
+```
````
Let's rename `config.py` to `run_group.py` or something similar.  Currently it could be mistaken for a tool that helps configure ONVM.  It is really more about running things than configuring them.

(`config.py` is a default name for configuring python packages)

twood02 commented on docs/NF_Dev.md:
```diff
@@ -99,17 +103,58 @@ values.
+      			"directory": "existing or new dir name"
+   		}
+	],
+	"NF name": [
```
have you tested that the program fails cleanly with a useful error if you give it a junk "NF Name" parameter?

catherinemeadows replied:
Current behavior is that it will exit and fail with the print statement: "Error: Unable to access NF directory <NF NAME>. Check syntax in your configuration file"

twood02 replied:
Ok that's fine.

EthanBaron14 commented on examples/run_group.py:
```diff
@@ -0,0 +1,218 @@
+        start = time.time()
+        time.clock()
+        elapsed = 0
+        while elapsed < timeout:
```
This timeout variable is not checked for proper type. If someone makes a string, say `"10"`, instead of `10`, they'll get a nasty error message of a Python TypeError. Maybe there's a way to check this input? Seeing this with most inputs. We don't really have a check for the `directory` field either, or the parameters for NFs. It just takes the input and as long as it's reasonably correct JSON, the program tries to continue. However, it realizes quickly that the types provided do not match what's expected and crashes.
